url3aa01d34ec_katsuster_ememu_tgz-pJ8-net_katsuster_ememu_ui_MainJ8/src/net/katsuster/ememu/arm/core/ExecStageARMv5.java

```diff
@@ -0,0 +1,4246 @@
+     * ビット[4]: 0
+     *
+     * 該当するオペランド:
+     * 数値はビット[6:4] の値を示す。
+     * 0b000: データ処理オペランド - レジスタ
+     * 0b000: データ処理オペランド - イミディエート論理左シフト
+     * 0b010: データ処理オペランド - イミディエート論理右シフト
+     * 0b100: データ処理オペランド - イミディエート算術右シフト
+     * 0b110: データ処理オペランド - イミディエート右ローテート
+     * 0b110: データ処理オペランド - 拡張付き右ローテート
+     *
+     * @param inst ARM 命令
+     * @return キャリーアウトする場合は true、そうでなければ false
+     */
+    public boolean getAddrMode1CarryImmShift(InstructionARM inst) {
+        int shift_imm = inst.getField(7, 5);
+        int shift = inst.getField(5, 2);
+        int rm = inst.getRmField();
+
+        switch (shift) {
+        case 0:
+            if (shift_imm == 0) {
+                //レジスタ
+                return getCPSR().getCBit();
+            } else {
+                //イミディエート論理左シフト
+                return BitOp.getBit32(getReg(rm), 32 - shift_imm);
+            }
+        case 1:
+            //イミディエート論理右シフト
+            if (shift_imm == 0) {
+                return BitOp.getBit32(getReg(rm), 31);
+            } else {
+                return BitOp.getBit32(getReg(rm), shift_imm - 1);
+            }
+        case 2:
+            //イミディエート算術右シフト
+            if (shift_imm == 0) {
+                return BitOp.getBit32(getReg(rm), 31);
+            } else {
+                return BitOp.getBit32(getReg(rm), shift_imm - 1);
+            }
+        case 3:
+            if (shift_imm == 0) {
+                //拡張付き右ローテート
+                return BitOp.getBit32(getReg(rm), 0);
+            } else {
+                //イミディエート右ローテート
+                return BitOp.getBit32(getReg(rm), shift_imm - 1);
+            }
+        default:
+            //do nothing
+            break;
+        }
+
+        throw new IllegalArgumentException("Unknown Imm Shift " +
+                String.format("0x%08x, shift:%d.",
+                        inst.getInst(), shift));
+    }
+
+    /**
+     * アドレシングモード 1 - データ処理オペランド、
+     * レジスタシフトのキャリーアウトを取得します。
+     *
+     * 条件:
+     * I ビット（ビット[25]）: 0
+     * ビット[4]: 1
+     *
+     * 該当するオペランド:
+     * 数値はビット[6:4] の値を示す。
+     * 0b001: データ処理オペランド - レジスタ論理左シフト
+     * 0b011: データ処理オペランド - レジスタ論理右シフト
+     * 0b101: データ処理オペランド - レジスタ算術右シフト
+     * 0b111: データ処理オペランド - レジスタ右ローテート
+     *
+     * @param inst ARM 命令
+     * @return キャリーアウトする場合は true、そうでなければ false
+     */
+    public boolean getAddrMode1CarryRegShift(InstructionARM inst) {
+        int shift = inst.getField(5, 2);
+        int rs = inst.getField(8, 4);
+        int rm = inst.getRmField();
+        int valRs, valRsLow;
+
+        //Rs[7:0]
+        valRs = getReg(rs) & 0xff;
+        //Rs[4:0]
+        valRsLow = getReg(rs) & 0x1f;
+
+        switch (shift) {
+        case 0:
+            //レジスタ論理左シフト
+            if (valRs == 0) {
+                return getCPSR().getCBit();
+            } else if (valRs <= 32) {
+                return BitOp.getBit32(getReg(rm), 32 - valRs);
+            } else {
+                return false;
+            }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 452:458, 461:467.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.